libs/ml/benchmark/embeddings/embeddings.cpp

```diff
@@ -0,0 +1,204 @@
+//------------------------------------------------------------------------------
+//
+//   Copyright 2018-2019 Fetch.AI Limited
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//
+//------------------------------------------------------------------------------
+
+#include "logging/logging.hpp"
+#include "math/tensor.hpp"
+#include "ml/core/graph.hpp"
+#include "ml/layers/fully_connected.hpp"
+#include "ml/ops/activations/relu.hpp"
+#include "ml/ops/loss_functions/mean_square_error_loss.hpp"
+#include "ml/optimisation/adam_optimiser.hpp"
+#include "ml/optimisation/lazy_adam_optimiser.hpp"
+
+#include "benchmark/benchmark.h"
+
+#include <memory>
+#include <string>
+
+//////////////////////////
+/// reusable functions ///
+//////////////////////////
+
+template <typename TypeParam>
+std::shared_ptr<fetch::ml::Graph<TypeParam>> PrepareTestGraph(
+    typename TypeParam::SizeType embedding_dimensions, typename TypeParam::SizeType n_datapoints,
+    std::string &input_name, std::string &label_name, std::string &error_name)
+{
+  std::shared_ptr<fetch::ml::Graph<TypeParam>> g(std::make_shared<fetch::ml::Graph<TypeParam>>());
+
+  input_name = g->template AddNode<fetch::ml::ops::PlaceHolder<TypeParam>>("", {});
+
+  std::string output_name = g->template AddNode<fetch::ml::ops::Embeddings<TypeParam>>(
+      "Embeddings", {input_name}, embedding_dimensions, n_datapoints);
+
+  label_name = g->template AddNode<fetch::ml::ops::PlaceHolder<TypeParam>>("", {});
+  error_name = g->template AddNode<fetch::ml::ops::MeanSquareErrorLoss<TypeParam>>(
+      "Error", {output_name, label_name});
+
+  return g;
+}
+
+template <typename T, fetch::math::SizeType B, fetch::math::SizeType S, fetch::math::SizeType D,
+          fetch::math::SizeType E>
+void BM_Setup_And_Train_Embeddings_Adam(benchmark::State &state)
```

Comment:
please add a function comment here explaining the point of this benchmark. Is it to get an idea of the overall time cost of training a simple embeddings model under different hyperparameters?